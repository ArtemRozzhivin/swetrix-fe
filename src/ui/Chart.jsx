import React, { useEffect } from "react";
import bb from "billboard.js";
import "billboard.js/dist/theme/datalab.css"
import "billboard.js/dist/billboard.css";

const Chart = ({current, children, options}) => {
  let chartInstance = {};

  useEffect(() => {
    renderChart()

    return () => { destroy() }
  })

  const destroy = () => {
    if (chartInstance !== null) {
      try {
        chartInstance.destroy();
      } catch (error) {
        console.error("Internal billboard.js error", error);
      } finally {
        chartInstance = null;
      }
    }
  }

  const renderChart = () => {
    if (current !== null) {
      chartInstance = bb.generate({
        ...options,
        bindto: "#"+ current
      });
    }
  }

  
  return (
    <div>
      <div id={current} />
      {children}
    </div>
  );

}

export default Chart
