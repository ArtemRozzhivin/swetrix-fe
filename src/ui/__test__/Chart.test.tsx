/* eslint-disable testing-library/prefer-screen-queries */
/* eslint-disable import/no-extraneous-dependencies */
import React from 'react'
import { render, cleanup } from '@testing-library/react'
import Chart from '../Chart'

describe('Chart component', () => {
  afterEach(cleanup)

  const options = {
    data: {
      columns: [
        ['data1', 30, 200, 100, 400, 150, 250],
        ['data2', 50, 20, 10, 40, 15, 25],
      ],
    },
  }

  it('should render the chart', () => {
    const { getByTestId } = render(<Chart current='chart' options={options} />)
    expect(getByTestId('chart')).toBeInTheDocument()
  })

  it('snapshot', () => {
    const { container } = render(<Chart current='chart' options={options} />)
    expect(container).toMatchSnapshot()
  })
})
