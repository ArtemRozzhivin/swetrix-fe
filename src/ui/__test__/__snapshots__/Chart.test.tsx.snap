// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart component snapshot 1`] = `
<div>
  <div>
    <div
      class="bb"
      data-testid="chart"
      id="chart"
      style="position: relative;"
    >
      <svg
        height="320"
        style="overflow: hidden; display: block;"
        width="0"
      >
        <defs>
          <clippath
            id="bb-8754287941272330-clip"
          >
            <rect
              height="226"
              width="0"
            />
          </clippath>
          <clippath
            id="bb-8754287941272330-clip-xaxis"
          >
            <rect
              height="110"
              width="20"
              x="-9"
              y="-15"
            />
          </clippath>
          <clippath
            id="bb-8754287941272330-clip-yaxis"
          >
            <rect
              height="240"
              width="50"
              x="-29"
              y="-4"
            />
          </clippath>
          <clippath
            id="bb-8754287941272330-clip-grid"
          >
            <rect
              height="226"
              width="0"
            />
          </clippath>
        </defs>
        <g
          class="bb-main"
          transform="translate(30.5, 4.5)"
        >
          <g
            class="bb-grid"
            clip-path="url(http://localhost/#bb-8754287941272330-clip-grid)"
          >
            <g
              class="bb-xgrid-focus"
            >
              <line
                class="bb-xgrid-focus"
                x1="-10"
                x2="-10"
                y1="0"
                y2="226"
              />
            </g>
          </g>
          <g
            class="bb-chart"
            clip-path="url(http://localhost/#bb-8754287941272330-clip)"
          >
            <g
              class="bb-event-rects bb-event-rects-single"
              style="fill-opacity: 0;"
            >
              <rect
                class="bb-event-rect"
                height="226"
                width="0"
                x="0"
                y="0"
              />
            </g>
            <g
              class="bb-chart-lines"
            >
              <g
                class="bb-chart-line bb-target bb-target-data1"
                style="pointer-events: none;"
              >
                <g
                  class="bb-shapes bb-shapes-data1 bb-lines bb-lines-data1"
                >
                  <path
                    class="bb-shape bb-shape bb-line bb-line-data1 "
                    d="M0,197.635L0,115.904L0,163.981L0,19.75L0,139.942L0,91.865"
                    style="stroke: #1f77b4;"
                  />
                </g>
                <g
                  class="bb-shapes bb-shapes-data1 bb-circles bb-circles-data1"
                >
                  <circle
                    class="bb-shape bb-shape-0 bb-circle bb-circle-0"
                    cx="0"
                    cy="197.6346153846154"
                    r="2.5"
                    style="fill: #1f77b4; stroke: #1f77b4;"
                  />
                  <circle
                    class="bb-shape bb-shape-1 bb-circle bb-circle-1"
                    cx="0"
                    cy="115.90384615384615"
                    r="2.5"
                    style="fill: #1f77b4; stroke: #1f77b4;"
                  />
                  <circle
                    class="bb-shape bb-shape-2 bb-circle bb-circle-2"
                    cx="0"
                    cy="163.98076923076923"
                    r="2.5"
                    style="fill: #1f77b4; stroke: #1f77b4;"
                  />
                  <circle
                    class="bb-shape bb-shape-3 bb-circle bb-circle-3"
                    cx="0"
                    cy="19.75000000000001"
                    r="2.5"
                    style="fill: #1f77b4; stroke: #1f77b4;"
                  />
                  <circle
                    class="bb-shape bb-shape-4 bb-circle bb-circle-4"
                    cx="0"
                    cy="139.9423076923077"
                    r="2.5"
                    style="fill: #1f77b4; stroke: #1f77b4;"
                  />
                  <circle
                    class="bb-shape bb-shape-5 bb-circle bb-circle-5"
                    cx="0"
                    cy="91.86538461538461"
                    r="2.5"
                    style="fill: #1f77b4; stroke: #1f77b4;"
                  />
                </g>
              </g>
              <g
                class="bb-chart-line bb-target bb-target-data2"
                style="pointer-events: none;"
              >
                <g
                  class="bb-shapes bb-shapes-data2 bb-lines bb-lines-data2"
                >
                  <path
                    class="bb-shape bb-shape bb-line bb-line-data2 "
                    d="M0,188.019L0,202.442L0,207.25L0,192.827L0,204.846L0,200.038"
                    style="stroke: #ff7f0e;"
                  />
                </g>
                <g
                  class="bb-shapes bb-shapes-data2 bb-circles bb-circles-data2"
                >
                  <circle
                    class="bb-shape bb-shape-0 bb-circle bb-circle-0"
                    cx="0"
                    cy="188.01923076923075"
                    r="2.5"
                    style="fill: #ff7f0e; stroke: #ff7f0e;"
                  />
                  <circle
                    class="bb-shape bb-shape-1 bb-circle bb-circle-1"
                    cx="0"
                    cy="202.4423076923077"
                    r="2.5"
                    style="fill: #ff7f0e; stroke: #ff7f0e;"
                  />
                  <circle
                    class="bb-shape bb-shape-2 bb-circle bb-circle-2"
                    cx="0"
                    cy="207.25"
                    r="2.5"
                    style="fill: #ff7f0e; stroke: #ff7f0e;"
                  />
                  <circle
                    class="bb-shape bb-shape-3 bb-circle bb-circle-3"
                    cx="0"
                    cy="192.82692307692307"
                    r="2.5"
                    style="fill: #ff7f0e; stroke: #ff7f0e;"
                  />
                  <circle
                    class="bb-shape bb-shape-4 bb-circle bb-circle-4"
                    cx="0"
                    cy="204.84615384615384"
                    r="2.5"
                    style="fill: #ff7f0e; stroke: #ff7f0e;"
                  />
                  <circle
                    class="bb-shape bb-shape-5 bb-circle bb-circle-5"
                    cx="0"
                    cy="200.03846153846152"
                    r="2.5"
                    style="fill: #ff7f0e; stroke: #ff7f0e;"
                  />
                </g>
              </g>
            </g>
          </g>
          <g
            class="bb-axis bb-axis-x"
            clip-path="url(http://localhost/#bb-8754287941272330-clip-xaxis)"
            transform="translate(0, 226)"
          >
            <text
              class="bb-axis-x-label"
              dx="-0.5em"
              dy="-0.5em"
              style="text-anchor: end;"
              x="0"
            />
            <g
              class="tick"
              transform="translate(0,0)"
            >
              <line
                x1="0"
                x2="0"
                y2="0"
              />
              <text
                style="text-anchor: middle;"
                x="0"
                y="9"
              >
                <tspan
                  dx="0"
                  dy=".71em"
                  x="0"
                >
                  0
                </tspan>
              </text>
            </g>
            <g
              class="tick"
              transform="translate(0,0)"
            >
              <line
                x1="0"
                x2="0"
                y2="0"
              />
              <text
                style="text-anchor: middle;"
                x="0"
                y="9"
              >
                <tspan
                  dx="0"
                  dy=".71em"
                  x="0"
                >
                  1
                </tspan>
              </text>
            </g>
            <g
              class="tick"
              transform="translate(0,0)"
            >
              <line
                x1="0"
                x2="0"
                y2="0"
              />
              <text
                style="text-anchor: middle;"
                x="0"
                y="9"
              >
                <tspan
                  dx="0"
                  dy=".71em"
                  x="0"
                >
                  2
                </tspan>
              </text>
            </g>
            <g
              class="tick"
              transform="translate(0,0)"
            >
              <line
                x1="0"
                x2="0"
                y2="0"
              />
              <text
                style="text-anchor: middle;"
                x="0"
                y="9"
              >
                <tspan
                  dx="0"
                  dy=".71em"
                  x="0"
                >
                  3
                </tspan>
              </text>
            </g>
            <g
              class="tick"
              transform="translate(0,0)"
            >
              <line
                x1="0"
                x2="0"
                y2="0"
              />
              <text
                style="text-anchor: middle;"
                x="0"
                y="9"
              >
                <tspan
                  dx="0"
                  dy=".71em"
                  x="0"
                >
                  4
                </tspan>
              </text>
            </g>
            <g
              class="tick"
              transform="translate(0,0)"
            >
              <line
                x1="0"
                x2="0"
                y2="0"
              />
              <text
                style="text-anchor: middle;"
                x="0"
                y="9"
              >
                <tspan
                  dx="0"
                  dy=".71em"
                  x="0"
                >
                  5
                </tspan>
              </text>
            </g>
            <path
              class="domain"
              d="M0,6V0H0V6"
            />
          </g>
          <g
            class="bb-axis bb-axis-y"
            clip-path="url(http://localhost/#bb-8754287941272330-clip-yaxis)"
            transform="translate(0, 0)"
          >
            <text
              class="bb-axis-y-label"
              dx="-0.5em"
              dy="1.2em"
              style="text-anchor: end;"
              transform="rotate(-90)"
              x="0"
            />
            <g
              class="tick"
              transform="translate(0,213)"
            >
              <line
                x2="-6"
                y1="0"
                y2="0"
              />
              <text
                style="text-anchor: end;"
                x="-9"
                y="0"
              >
                <tspan
                  dx="0"
                  dy="3"
                  x="-9"
                >
                  0
                </tspan>
              </text>
            </g>
            <g
              class="tick"
              transform="translate(0,189)"
            >
              <line
                x2="-6"
                y1="0"
                y2="0"
              />
              <text
                style="text-anchor: end;"
                x="-9"
                y="0"
              >
                <tspan
                  dx="0"
                  dy="3"
                  x="-9"
                >
                  50
                </tspan>
              </text>
            </g>
            <g
              class="tick"
              transform="translate(0,164)"
            >
              <line
                x2="-6"
                y1="0"
                y2="0"
              />
              <text
                style="text-anchor: end;"
                x="-9"
                y="0"
              >
                <tspan
                  dx="0"
                  dy="3"
                  x="-9"
                >
                  100
                </tspan>
              </text>
            </g>
            <g
              class="tick"
              transform="translate(0,140)"
            >
              <line
                x2="-6"
                y1="0"
                y2="0"
              />
              <text
                style="text-anchor: end;"
                x="-9"
                y="0"
              >
                <tspan
                  dx="0"
                  dy="3"
                  x="-9"
                >
                  150
                </tspan>
              </text>
            </g>
            <g
              class="tick"
              transform="translate(0,116)"
            >
              <line
                x2="-6"
                y1="0"
                y2="0"
              />
              <text
                style="text-anchor: end;"
                x="-9"
                y="0"
              >
                <tspan
                  dx="0"
                  dy="3"
                  x="-9"
                >
                  200
                </tspan>
              </text>
            </g>
            <g
              class="tick"
              transform="translate(0,92)"
            >
              <line
                x2="-6"
                y1="0"
                y2="0"
              />
              <text
                style="text-anchor: end;"
                x="-9"
                y="0"
              >
                <tspan
                  dx="0"
                  dy="3"
                  x="-9"
                >
                  250
                </tspan>
              </text>
            </g>
            <g
              class="tick"
              transform="translate(0,68)"
            >
              <line
                x2="-6"
                y1="0"
                y2="0"
              />
              <text
                style="text-anchor: end;"
                x="-9"
                y="0"
              >
                <tspan
                  dx="0"
                  dy="3"
                  x="-9"
                >
                  300
                </tspan>
              </text>
            </g>
            <g
              class="tick"
              transform="translate(0,44)"
            >
              <line
                x2="-6"
                y1="0"
                y2="0"
              />
              <text
                style="text-anchor: end;"
                x="-9"
                y="0"
              >
                <tspan
                  dx="0"
                  dy="3"
                  x="-9"
                >
                  350
                </tspan>
              </text>
            </g>
            <g
              class="tick"
              transform="translate(0,20)"
            >
              <line
                x2="-6"
                y1="0"
                y2="0"
              />
              <text
                style="text-anchor: end;"
                x="-9"
                y="0"
              >
                <tspan
                  dx="0"
                  dy="3"
                  x="-9"
                >
                  400
                </tspan>
              </text>
            </g>
            <path
              class="domain"
              d="M-6,1H0V226H-6"
            />
          </g>
        </g>
        <g
          class="bb-legend"
          transform="translate(0, 260)"
        >
          <g
            class="bb-legend-item bb-legend-item-data1"
            style="cursor: pointer;"
          >
            <text
              style="pointer-events: none;"
              x="1.5"
              y="13"
            >
              data1
            </text>
            <rect
              class="bb-legend-item-event"
              height="4"
              style="fill-opacity: 0;"
              width="25"
              x="-12.5"
              y="-1"
            />
            <line
              class="bb-legend-item-tile"
              stroke-width="10"
              style="stroke: #1f77b4; pointer-events: none;"
              x1="-14.5"
              x2="-4.5"
              y1="8"
              y2="8"
            />
          </g>
          <g
            class="bb-legend-item bb-legend-item-data2"
            style="cursor: pointer;"
          >
            <text
              style="pointer-events: none;"
              x="6.5"
              y="17"
            >
              data2
            </text>
            <rect
              class="bb-legend-item-event"
              height="4"
              style="fill-opacity: 0;"
              width="15"
              x="-7.5"
              y="3"
            />
            <line
              class="bb-legend-item-tile"
              stroke-width="10"
              style="stroke: #ff7f0e; pointer-events: none;"
              x1="-9.5"
              x2="0.5"
              y1="12"
              y2="12"
            />
          </g>
        </g>
      </svg>
      <div
        class="bb-tooltip-container"
        style="position: absolute; pointer-events: none; display: none;"
      />
    </div>
  </div>
</div>
`;
